doc/sphinx/source/driver_design.rst

```diff
@@ -0,0 +1,179 @@
+
+Setting device properties
+-----------------------------------
+
+In the driver's init callback, the driver should discover and set up each device's properties.
+These are stored as members of ``struct _cl_device_id``, e.g. ``max_work_item_dimensions`` or
+``max_compute_units``. Some of them are internal to PoCL, e.g. ``device_side_printf`` is a bool
+that changes the processing of the 'printf' function in PoCL's LLVM passes. There are some helpers
+for setting up properties; for setting up versioned (OpenCL 3.0) properties, see ``lib/CL/devices/common.c``;
+the same file also has ``pocl_init_default_device_infos`` which sets up some reasonable defaults.
+By default the entire struct is memset to zero, so all properties are zero / false.
+
+
+PoCL's internal driver API
+-----------------------------------
+
+Next step is to implement PoCL's internal driver API operations (callbacks) for the device.
+They're described in lib/CL/pocl_cl.h, in ``struct pocl_device_ops``. The declarations for all
+operations can be obtained by including ``prototypes.inc`` and using the ``GEN_PROTOTYPES`` macro.
+The declarations will be named ``pocl_<DRIVER_NAME>_<OP_NAME>``, e.g. ``pocl_level0_map_mem``.
+Many of these callbacks are optional; there's no need to implement all to get a useful driver.
+Any callback that is implemented, needs to be assigned in the ``ops`` struct of the
+``struct _cl_device_id`` in the ``pocl_<DRIVER>_init_device_ops`` function. For an example, see
+``pocl_basic_init_device_ops`` in ``lib/CL/devices/basic/basic.c``. If the callback is optional
+and will not be implemented, it can be left as NULL in ``ops``.
+
+The callbacks can be divided into several groups:
+
+* probe & initialization (required)
+* memory mapping, allocation & freeing (required)
+* event handling (required)
+* compilation callbacks (required)
+* command execution callbacks (optional)
+* SVM callbacks (optional)
+* Image callbacks (optional)
+* Object create / free callbacks (optional)
+* D2D migration callbacks (optional)
+
+1. probe & initialization - ``probe, init, reinit, uninit``
+
+   you'll need to implement at least ``probe`` and ``init``; ``probe`` is only called once, when
+   libpocl is loaded, and its job is to find out how many devices are available.
+   ``init`` is then called for each device in succession. Implementing ``reinit`` & ``uninit``
+   is optional - these release any resources after the last context is released.
+   This is useful mainly for debugging memory leaks; since there is no ``clReleasePlatform``
+   in OpenCL, calling uninit should release everything - anything left over is a leak.
+
+2. memory allocation - ``alloc_mem_obj, free``
+
+   these allocate & free device-side memory. See documentation on ``struct _cl_mem->device_ptrs``
+   and ``pocl_mem_identifier``. If the device supports images, these must also allocate/free images.
+   Memory allocation in PoCL is done lazily, IOW a buffer is allocated only on devices will run
```

Comment:
But still "actively" timing-wise, right? That is, we reserve the space immediately, not delay the allocation to when it's needed (before the first kernel that uses it is launched)? Perhaps clarify that.